src/python/pants/backend/project_info/peek.py

```diff
@@ -0,0 +1,159 @@
+from enum import Enum
+from typing import Iterable, cast
+
+from pkg_resources import Requirement
+
+from pants.engine.addresses import Address, Addresses, BuildFileAddress
+from pants.engine.console import Console
+from pants.engine.fs import DigestContents, FileContent, PathGlobs
+from pants.engine.goal import Goal, GoalSubsystem, Outputting
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import Target, UnexpandedTargets
+
+
+class OutputOptions(Enum):
+    RAW = "raw"
+    JSON = "json"
+
+
+class PeekSubsystem(Outputting, GoalSubsystem):
+    """Display BUILD file info to the console.
+
+    In its most basic form, `peek` just prints the contents of a BUILD file. It can also display
+    multiple BUILD files, or render normalized target metadata as JSON for consumption by other
+    programs.
+    """
+
+    name = "peek"
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--output",
+            type=OutputOptions,
+            default=OutputOptions.JSON,
+            help="which output style peek should use",
+        )
+        register(
+            "--exclude-defaults",
+            type=bool,
+            default=False,
+            help="whether to exclude values matching default field values from normalized output",
+        )
+
+    @property
+    def output_type(self) -> OutputOptions:
+        """Get the output type from options.
+
+        Must be renamed here because `output` conflicts with `Outputting` class.
+        """
+        return cast(OutputOptions, self.options.output)
+
+    @property
+    def exclude_defaults(self) -> bool:
+        return cast(bool, self.options.exclude_defaults)
+
+
+class Peek(Goal):
+    subsystem_cls = PeekSubsystem
+
+
+def _render_raw(fcs: Iterable[FileContent]) -> str:
+    sorted_fcs = sorted(fcs, key=lambda fc: fc.path)
+    rendereds = map(_render_raw_build_file, sorted_fcs)
+    return os.linesep.join(rendereds)
+
+
+def _render_raw_build_file(fc: FileContent, encoding: str = "utf-8") -> str:
+    dashes = "-" * len(fc.path)
+    content = fc.content.decode(encoding)
+    parts = [dashes, fc.path, dashes, content]
+    if not content.endswith(os.linesep):
+        parts.append("")
+    return os.linesep.join(parts)
+
+
+def _render_json(ts: Iterable[Target], exclude_defaults: bool = False) -> str:
+    targets = [_target_to_dict(t, exclude_defaults) for t in ts]
+    data = dict(targets=targets)
+    return json.dumps(data, indent=2, cls=_PeekJsonEncoder)
+
+
+_nothing = object()
+
+
+def _target_to_dict(t: Target, exclude_defaults: bool = False, _nothing: object = _nothing) -> dict:
+    return {
+        "address": t.address.spec,
+        "target_type": t.alias,
+        **{
+            k.alias: v.value
+            for k, v in t.field_values.items()
+            # don't report default values in normalized output
```

Comment:
Nit: comments should be sentences: https://www.pantsbuild.org/docs/style-guide#style.